Openmp profiling

pp1336 opened the pull request:
An addition to PR#89, allows profiling under OpenMP settings.

vincepandolfo commented on devito/propagator.py:
```diff
@@ -308,7 +308,11 @@ def generate_loops(self, loop_body):
         omp_single = [cgen.Pragma("omp single")] if self.compiler.openmp else []
         omp_parallel = [cgen.Pragma("omp parallel")] if self.compiler.openmp else []
-        omp_for = [cgen.Pragma("omp for schedule(static)")] if self.compiler.openmp else []
+        reduction_list = ["kernel", "loop_body"] if self.compiler.openmp else []
```
This should be managed by the Profiler instead of being hard coded.

pp1336 replied:
Well as we discussed just now, the other parts of the profiling code are also hard coding the these stuff, so I think this is a more general issue that deserves another PR.

vincepandolfo commented on devito/propagator.py:
```diff
@@ -308,7 +308,11 @@ def generate_loops(self, loop_body):
+        omp_for = [cgen.Pragma("omp for schedule(static)"\
+                   + (self.profiler.get_loop_reduction("+", reduction_list)\
+                   if self.profile else ""))]\
+                  if self.compiler.openmp else []
```
This should be managed by the Profiler as well, so that we can do a different reduction on timing since that doesn't work with OpenMP and `+` is not the right reduction here I believe.

pp1336 replied:
the + is only for flops, I did know that timing was not ok with openmp guess a little miscommunication here, I will add the change to handle loop timing

mlange05 commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+        if len(variables) <= 0:
+            return ""
+        return " reduction(" + op + ":" + self.loop_temp_prefix + variables[0]\
+               + "".join(["," + self.loop_temp_prefix + v for v in variables[1:]]) + ")"
```
Using `+` to concatenate strings is considered bad style by many. String formatting (`%s`) is much cleaner and more readable.

mlange05 commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+        if len(variables) <= 0:
+            return ""
+        return "long long " + self.loop_temp_prefix + variables[0] + " = " + value +\
+               "".join([", " + self.loop_temp_prefix + v + " = " + value for v in variables[1:]])
```
String formatting

mlange05 commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+        :returns: A list of string representing statements used as updating
+                  statements
+        """
+        return [self.f_name + "->" + v + " += " + self.loop_temp_prefix + v for v in variables]
```
String formatting

mlange05 commented:
Ok, the code looks great, except for a few minor string formatting comments. One thing I found when running this though is that `gflops` seem to double with two threads, which I don't think is correct. OI and timings look sensible though.

FabioLuporini commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+        self.t_name = "timings"
+        self.o_name = "oi"
+        self.f_name = "flops"
+        self.loop_temp_prefix = "temp_"
```
Feels like ``t_name, o_name, ..., loop_temp_prefix`` should actually be class (read-only) attributes ?
Also, the names you chose are not very explanatory

vincepandolfo replied:
What would better names be? I agree on having the as read-only attributes.

FabioLuporini replied:
Maybe ``_C_timings``, ``_C_oi``, ``_C_flops``, ``_C_temporary``. Plus add a comment that says that attributes starting with ``_C_`` are simply used for code generation ?

FabioLuporini commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+    def get_loop_flop_update(self, variables):
+        """Function to generate a list of statement representing the amount
+        of flops update per iteration of the loop.
+        :param variables: A list of string, each is the suffix of the
```
Leave a blank line here before ``:param variables: ...``

FabioLuporini commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+    def get_loop_temp_var_decl(self, value, variables):
+        """Function to generate the declariation and initialisation of the
+        temporary variables used for reduction for loop profiling.
+        :param value: String represeting the initial value of temporary
```
Leave a blank line here before ``:param value: ...``

FabioLuporini commented on devito/profiler.py:
```diff
@@ -1,30 +1,86 @@
+    def get_loop_reduction(self, op, variables):
+        """Function to generate reduction pragma, used for profiling under
+        OpenMP settings.
+        :param op: The reduction operator.
```
Leave a blank line here before ``:param op: ...``

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+        :param code: The code to be profiled.
+        """
+        if name not in self.oi:
+            self.oi[name] = 0
```
Consider using a ``defaultdict(int)`` instead of a ``{}`` ?

``from collections import defaultdict``

vincepandolfo replied:
Why? I'm not familiar with `defaultdict`

FabioLuporini replied:
If you use a ``defaultdict``, an entry which is not in the dict gets assigned a default value (0 in the case of int). Look at this:

```
In [1]: from collections import defaultdict

In [2]: d = defaultdict(int)

In [3]: d[1]
Out[3]: 0

In [4]: d[2] += 1

In [5]: d
Out[5]: defaultdict(<type 'int'>, {1: 0, 2: 1})
```

Similar behaviour with a normal dict and using the ``setdefault`` method ({}.setdefault.__doc__ is quite self-explanatory)

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+        if name not in self.oi:
+            self.oi[name] = 0
+        if name not in self.flops_defaults:
+            self.flops_defaults[name] = 0
```
same concept as above

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+            elif isinstance(elem, Block):
+                block_oi, block_flops = self._get_block_oi_and_flops(name, elem, loads)
+                self.oi[name] += block_oi
+                self.flops_defaults[name] += block_flops
```
What about AugmentedAssignments, such as Incr, Decr (i.e., +=, -=) ... ?

Maybe we should also have a default ``else`` branch which prints a warning if hit ?

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+                    idx += 1
+                else:
+                    cur_load += char
+                    idx += 1
```
Ah, maybe I see now. This looks very hacky. I think we should make a better use of the cgen hierarchy (by suitably handling Assign, Incr, Decr etc)

vincepandolfo replied:
The only `cgen` objects we use now are `For`, `Block` and `Assign`. That's why it is structured this way.

FabioLuporini replied:
The method is still hacky, though

The right way of doing this, I think, is to have a visitor that traverses cgen expressions, not analysing strings textually

vincepandolfo replied:
The string is the only thing I get from cgen. There's no expression.

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+            return {}
+
+        return dict((field, getattr(self._timings, field))
+                    for field, _ in self._timings._fields_)
```
I'm not sure I understand this, maybe add an example as a comment (perhaps as method doc ?)

vincepandolfo replied:
What's that you don't get?

FabioLuporini replied:
ah, the content of the resulting dict

vincepandolfo replied:
`self._timings` is a ctypes struct. `self._timings._fields_` contains a list of tuples `(name, type)`. I get the name from there and the value with `getattr`.

It's simply a conversion from a ctypes struct to a `dict`

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+                     float(
+                         getattr(self._flops, field) + self.flops_defaults[field]
+                     )/self.timings[field]/10**9)
+                    for field, _ in self._flops._fields_)
```
Same as before

FabioLuporini commented on devito/profiler.py:
```diff
@@ -39,50 +95,208 @@ def add_profiling(self, block, name, omp_flag=None):
+        if self._flops:
+            return dict((field, int(getattr(self._flops, field))) for field, _ in self._flops._fields_)
+        else:
+            return {}
```
Looks more complicated than it really needs to be ?

pp1336 replied:
I was just following the previous style, i dont mind changing it.

pp1336 commented:
I will handle the formatting issue, @vincepandolfo do you want to handle the others since you probably know those code better

FabioLuporini commented:
@pp1336 we need to merge this ASAP. Please join slack.

pp1336 commented:
vincepandolfo and I have addressed most comments. We agree that we are good with the profiler as it is now, and that the profiler might require further improvements in the future, when we have time.

pp1336 commented:
OK so we are just waiting 81 to be merge, then we will rebase this branch and merge.

pp1336 commented:
Looks like all tests under Linux passed, failed due to flake8. All tests for mac failed.
